### DISCUSSION TITLE
Getting origin from student-placed pl-coordinates element that is rotated

### DISCUSSION THREAD

**Discussion by ffund** [PrairieLearn/PrairieLearn#6101] [Q&A] [answered]
I cannot seem to figure out the geometry of the `pl-coordinates` part. I am trying to get the origin of the student-placed coordinate system in a `grade` function. In a `pl-coordinates`, the origin is not exactly in the corner of the student-placed object.

(The `x1` and `y1` HTML attributes are the origin of the coordinate system.) If there is no rotation, then `left` and `top` give the x and y values, respectively, of the origin of the coordinate system. But if the placed element is rotated, then I'm not sure what `left` and `top` are.

Example:  if I have this coordinate system:

```
<pl-coordinates x1="170.80" y1="247.91" width="70" label-x="PC1" label-y="PC2" angle="331.45"></pl-coordinates>
```

and I place a `pl-coordinates` as closely as I can on top of it, this is what I get in `grade`:

```
{'angle': 333.15417962913546,
  'arrowheadOffsetRatio': 1, 'arrowheadWidthRatio': 1, 
  'drawEndArrow': 1, 'drawStartArrow': 0, 'evented': 1, 'graded': 1, 
  'gradingName': 'pl-coordinates', 'id': 511, 'label': 0, 'labelx': 'PC1', 'labely': 'PC2',  
  'left': 139.6759920879159,  
  'offsetx': -16, 'offsetx_label_x': 0, 'offsetx_label_y': -20, 'offsety': -10, 
  'offsety_label_x': 0, 'offsety_label_y': -10, 'originY': 'center', 
  'placed_by_user': 1, 'selectable': 1, 'stroke': '#268bd2', 'strokeWidth': 2,  
  'top': 271.32254481513564, 
  'type': 'pl-coordinates', 'width': 70}
```

**Comment by echuber2** [comment, marked as answer]
@nicknytko @mfsilva22 ~May know more about the intended behavior.~ (See below for followup)

**Reply by echuber2** [reply]
I've been poking around at this and I believe there's a bug in `mechanicsObjects.js` causing this. I'll try to prepare a PR tomorrow.

**Reply by echuber2** [reply]
Here is a rough correction for now:

```python
import math

def correct_origin(left, top, angle, width):
    groupOffsetY = width + 2
    groupOffsetX = groupOffsetY / 2.0
    angle_rad = (math.pi / 180) * (360 - angle)
    correct_left = (left - groupOffsetX) + math.cos(angle_rad) * groupOffsetX + math.sin(angle_rad) * groupOffsetY
    correct_top = (top - groupOffsetY) + math.cos(angle_rad) * groupOffsetY - math.sin(angle_rad) * groupOffsetX
    return (correct_left, correct_top)
```

In your example:

```
>>> correct_origin(139.676, 271.323, 333.15, 70)
(168.31416587746338, 247.3011982987849)
```

and this roughly matches the `x1="170.80" y1="247.91"` that you expected.

I learned the correction factor from the implementation of `mechanicsObjects.byType['pl-coordinates']`. 

https://github.com/PrairieLearn/PrairieLearn/blob/c3e7dd2dbf525aae16e9e8613aa55e331b4aeabb/elements/pl-drawing/mechanicsObjects.js#L2330

The dynamic values of `groupOffsetX` and `groupOffsetY` depend on the `width`, but it's not obvious how exactly without experimentation, and that may be fragile in this post hoc correction. Then, the current definition of `update_labels` seems to do the right thing with its local updates to `x` and `y` for the labels even under rotation. In contrast, the definition of `modify` does _not_ make this rotational correction for the whole object's `left`/`top`. But it's not as simple as changing `modify`, because that seems to mess up the placement of the coordinates object after submission.

So, unfortunately, it looks like baking this fix into `mechanicsOjects.js` may break other stuff, so we may need to make a new property to track the corrected origin. I'm a bit mystified about how the bounding box is being calculated based on the width; the height can't be specified, and in my experiment, the labels are getting cut off. (If several different things are messed up with the coordinates object, then fixing any of those may require this correction factor to be reconsidered too.) Furthermore, it looks like grading a student-placed coordinates object doesn't work automatically, but perhaps that's exactly what you're trying to contribute.

**Reply by echuber2** [reply]
Made an issue here: https://github.com/PrairieLearn/PrairieLearn/issues/6104

**Reply by ffund** [reply]
Thank you!!!!! This correction is just what I needed, and I would never have figured it out myself.
